x/state-verifier/keeper/keeper.go

```diff
@@ -0,0 +1,158 @@
+
+	icqtypes "github.com/neutron-org/neutron/v5/x/interchainqueries/types"
+	"github.com/neutron-org/neutron/v5/x/state-verifier/types"
+)
+
+type (
+	Keeper struct {
+		cdc        codec.BinaryCodec
+		storeKey   storetypes.StoreKey
+		cometInfo  comet.BlockInfoService
+		headerInfo header.Service
+		authority  string
+	}
+)
+
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	storeKey storetypes.StoreKey,
+	cometInfo comet.BlockInfoService,
+	headerInfo header.Service,
+	authority string,
+) *Keeper {
+	return &Keeper{
+		cdc:        cdc,
+		storeKey:   storeKey,
+		authority:  authority,
+		headerInfo: headerInfo,
+		cometInfo:  cometInfo,
+	}
+}
+
+func (k *Keeper) GetAuthority() string {
+	return k.authority
+}
+
+func (k *Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+// SaveConsensusState extracts info about the current header from the context, composes ConsensusState structure with that info
+// and saves the structure to the state
+func (k *Keeper) SaveConsensusState(ctx sdk.Context) error {
+	headerInfo := k.headerInfo.GetHeaderInfo(ctx)
+	cometInfo := k.cometInfo.GetCometBlockInfo(ctx)
+
+	cs := tendermint.ConsensusState{
+		Timestamp:          ctx.BlockTime(),
+		Root:               ibccommitmenttypes.NewMerkleRoot(headerInfo.AppHash),
+		NextValidatorsHash: cometInfo.GetValidatorsHash(),
+	}
+
+	return k.WriteConsensusState(ctx, ctx.BlockHeight(), cs)
+}
+
+// WriteConsensusState writes ConsensusState structure and corresponding height into the storage
+func (k *Keeper) WriteConsensusState(ctx sdk.Context, height int64, cs tendermint.ConsensusState) error {
+	store := ctx.KVStore(k.storeKey)
+	key := types.GetConsensusStateKey(height)
+
+	csBz, err := k.cdc.Marshal(&cs)
+	if err != nil {
+		return errors.Wrapf(sdkerrors.ErrJSONMarshal, err.Error())
+	}
+
+	store.Set(key, csBz)
+
+	return nil
+}
+
+// Verify verifies that provided `values` are actually present on Neutron blockchain at `blockHeight`
+func (k *Keeper) Verify(ctx sdk.Context, blockHeight int64, values []*icqtypes.StorageValue) error {
+	store := ctx.KVStore(k.storeKey)
+
+	// we need to use consensus state from the next height (N + 1), cause that consensus state contains .AppHash (Merkle Root) of the state for `blockHeight` (N)
+	csBz := store.Get(types.GetConsensusStateKey(blockHeight + 1))
+	if csBz == nil {
+		return errors.Wrap(sdkerrors.ErrKeyNotFound, fmt.Sprintf("consensus state for block %d not found", blockHeight))
+	}
+
+	var cs tendermint.ConsensusState
+	if err := k.cdc.Unmarshal(csBz, &cs); err != nil {
+		return errors.Wrap(sdkerrors.ErrJSONUnmarshal, err.Error())
+	}
+
+	for _, value := range values {
+		proof, err := ibccommitmenttypes.ConvertProofs(value.Proof)
+		if err != nil {
+			return errors.Wrapf(sdkerrors.ErrInvalidType, "failed to convert crypto.ProofOps to MerkleProof: %v", err)
+		}
+
+		path := ibccommitmenttypes.NewMerklePath(value.StoragePrefix, string(value.Key))
```

Comment:
The same code block is used in icq module https://github.com/neutron-org/neutron/blob/main/x/interchainqueries/keeper/msg_server.go#L254
How do you think to extract it into common method and reuse it, instead of just copying

Reply:
Done, please, check it